java/ql/src/experimental/Security/CWE/CWE-522/UnsecureBasicAuth.ql

```diff
@@ -0,0 +1,228 @@
+/**
+ * @name Unsecure basic authentication
+ * @description Basic authentication only obfuscates username/password in Base64 encoding, which can be easily recognized and reversed. Transmission of sensitive information not over HTTPS is vulnerable to packet sniffing.
+ * @kind problem
+ * @id java/unsecure-basic-auth
+ * @tags security
+ *       external/cwe-522
+ *       external/cwe-319
+ */
+
+import java
+import semmle.code.java.frameworks.Networking
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * The Java class `org.apache.http.client.methods.HttpRequestBase`. Popular subclasses include `HttpGet`, `HttpPost`, and `HttpPut`.
+ * And the Java class `org.apache.http.message.BasicHttpRequest`.
+ */
+class ApacheHttpRequest extends RefType {
+  ApacheHttpRequest() {
+    hasQualifiedName("org.apache.http.client.methods", "HttpRequestBase") or
+    hasQualifiedName("org.apache.http.message", "BasicHttpRequest")
+  }
+}
+
+/**
+ * Class of Java URL constructor.
+ */
+class URLConstructor extends ClassInstanceExpr {
+  URLConstructor() { this.getConstructor().getDeclaringType() instanceof TypeUrl }
+
+  Expr stringArg() {
+    // URLs constructed with any of the four string constructors below:
+    // URL(String spec), URL(String protocol, String host, int port, String file), URL(String protocol, String host, int port, String file, URLStreamHandler handler), URL(String protocol, String host, String file)
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    result = this.getArgument(0) // First argument contains the protocol part.
+  }
+}
+
+/**
+ * The type `java.net.URLConnection`.
+ */
+class TypeHttpUrlConnection extends RefType {
+  TypeHttpUrlConnection() { hasQualifiedName("java.net", "HttpURLConnection") }
```

Comment:
Consider adding `getASourceSupertype*()` to also include sub classes of `HttpURLConnection`.

Reply:
Thanks @aibaars. I've made the change.